lti/views.py

```diff
@@ -0,0 +1,183 @@
+        )
+
+        return super(LtiInitializerView, self).get(request, *args, **kwargs)
+
+    def post(self, request, *args, **kwargs):
+        """ Handle the POST coming directly from Canvas """
+
+        lti_params = {
+            "course_id": request.POST.get(LTI_PROPERTY_COURSE_ID, None),
+            "roles": request.POST.get(LTI_PROPERTY_USER_ROLES, None),
+            "title": request.POST.get(LTI_PROPERTY_COURSE_TITLE, None)
+        }
+
+        if None in lti_params.values():
+            return render(request, "lti_failure.html")
+
+        self.initialize_group(
+            lti_params["course_id"],
+            lti_params["title"],
+            lti_params["roles"],
+            request.user
+        )
+
+        return super(LtiInitializerView, self).post(request, *args, **kwargs)
+
+    def initialize_group(self, course_id, title, roles, user):
+        """
+        Takes in (str)course_id (str)title (str)roles, and (obj)user.
+        Gets or creates group, and updates the user's role in that group.
+        The parameters cannot be None
+        Returns None
+        """
+        group = self.get_or_create_group(course_id=int(course_id), title=title, user=user)
+        # Determine if the user is a Teacher or Student
+        role = self.determine_role(roles)
+        # Update the roles at each launch
+        self.update_roles(user=user, group=group, role=role)
+        return None
+
+    def determine_role(self, roles):
+        """ Takes (list) roles or None, returns (str) role  """
+        if roles:
+            roles = [role.rsplit("/")[-1] for role in roles.split(",")]
+            if "Instructor" in roles:
+                return "Teacher"
+        # If roles is None, then I am returning "Student", although it is actually
+        # indicative on an LTI failure.
+        return "Student"
+
+    def get_or_create_group(self, course_id=None, title="Empty", user=None):
+        """ takes (int) course_id, and (str) title and returns a Group"""
+        try:
+            group = Group.objects.get(class_key=course_id)
+        except Group.DoesNotExist:
+            group = Group.objects.create(
+                class_key=course_id,
+                title=title,
+                created_by=user
+            )
+        return group
+
+    def update_roles(self, user, group, role):
+        """
+        Takes a user, group, and roles list.
+        Will update group's teachers or students sets
+        Returns None
+        """
+        if role == "Teacher":
+            group.teachers.add(user)
+            if user in group.students.all():
+                group.students.remove(user)
+        else:
+            group.students.add(user)
+            if user in group.teachers.all():
+                group.teachers.remove(user)
+        return None
+
+
+class LtiRegistrationView(FormView):
+
+    template_name = "lti_registration.html"
+    form_class = LtiUsernameForm
+    success_url = "/lti/lti_initializer/"
```

Comment:
By using `reverse` we can avoid hard coding the URL and keep things DRY:

```suggestion
    success_url = reverse("lti_initializer")
```

Reply:
This view has been removed